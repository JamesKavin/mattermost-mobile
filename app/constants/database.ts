// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import keyMirror from '@utils/key_mirror';

export const MM_TABLES = {
    APP: {
        INFO: 'Info',
        GLOBAL: 'Global',
        SERVERS: 'Servers',
    },
    SERVER: {
        CATEGORY: 'Category',
        CATEGORY_CHANNEL: 'CategoryChannel',
        CHANNEL: 'Channel',
        CHANNEL_INFO: 'ChannelInfo',
        CHANNEL_MEMBERSHIP: 'ChannelMembership',
        CONFIG: 'Config',
        CUSTOM_EMOJI: 'CustomEmoji',
        DRAFT: 'Draft',
        FILE: 'File',
        GROUP: 'Group',
        GROUP_CHANNEL: 'GroupChannel',
        GROUP_MEMBERSHIP: 'GroupMembership',
        GROUP_TEAM: 'GroupTeam',
        MY_CHANNEL: 'MyChannel',
        MY_CHANNEL_SETTINGS: 'MyChannelSettings',
        MY_TEAM: 'MyTeam',
        POST: 'Post',
        POSTS_IN_CHANNEL: 'PostsInChannel',
        POSTS_IN_THREAD: 'PostsInThread',
        PREFERENCE: 'Preference',
        REACTION: 'Reaction',
        ROLE: 'Role',
        SYSTEM: 'System',
        TEAM: 'Team',
        TEAM_CHANNEL_HISTORY: 'TeamChannelHistory',
        TEAM_MEMBERSHIP: 'TeamMembership',
        TEAM_SEARCH_HISTORY: 'TeamSearchHistory',
        THREAD: 'Thread',
        THREADS_IN_TEAM: 'ThreadsInTeam',
        THREAD_PARTICIPANT: 'ThreadParticipant',
        TEAM_THREADS_SYNC: 'TeamThreadsSync',
        USER: 'User',
    },
};

export const MIGRATION_EVENTS = keyMirror({
    MIGRATION_ERROR: null,
    MIGRATION_STARTED: null,
    MIGRATION_SUCCESS: null,
});

export const SYSTEM_IDENTIFIERS = {
    CONFIG: 'config',
    CURRENT_CHANNEL_ID: 'currentChannelId',
    LAST_UNREAD_CHANNEL_ID: 'lastUnreadChannelId',
    CURRENT_TEAM_ID: 'currentTeamId',
    CURRENT_USER_ID: 'currentUserId',
    DATA_RETENTION_POLICIES: 'dataRetentionPolicies',
    EXPANDED_LINKS: 'expandedLinks',
    GLOBAL_THREADS_TAB: 'globalThreadsTab',
    LAST_DISMISSED_BANNER: 'lastDismissedBanner',
    LAST_SERVER_VERSION_CHECK: 'LastServerVersionCheck',
    LICENSE: 'license',
    ONLY_UNREADS: 'onlyUnreads',
    PUSH_VERIFICATION_STATUS: 'pushVerificationStatus',
    RECENT_CUSTOM_STATUS: 'recentCustomStatus',
    RECENT_MENTIONS: 'recentMentions',
    RECENT_REACTIONS: 'recentReactions',
    SESSION_EXPIRATION: 'sessionExpiration',
    TEAM_HISTORY: 'teamHistory',
    WEBSOCKET: 'WebSocket',
};

export const GLOBAL_IDENTIFIERS = {
    DEVICE_TOKEN: 'deviceToken',
    DONT_ASK_FOR_REVIEW: 'dontAskForReview',
    FIRST_LAUNCH: 'firstLaunch',
    LAST_ASK_FOR_REVIEW: 'lastAskForReview',
    MULTI_SERVER_TUTORIAL: 'multiServerTutorial',
    PROFILE_LONG_PRESS_TUTORIAL: 'profileLongPressTutorial',
    ONBOARDING: 'onboarding',
};

export enum OperationType {
    CREATE = 'CREATE',
    UPDATE = 'UPDATE',
    DELETE = 'DELETE',
  }

// The only two types of databases in the app
export enum DatabaseType {
    DEFAULT,
    SERVER,
  }

export default {
    GLOBAL_IDENTIFIERS,
    MM_TABLES,
    MIGRATION_EVENTS,
    SYSTEM_IDENTIFIERS,
};
