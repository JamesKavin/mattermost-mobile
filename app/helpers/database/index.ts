// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {General} from '@constants';

import type Model from '@nozbe/watermelondb/Model';
import type ChannelModel from '@typings/database/models/servers/channel';

export const extractRecordsForTable = <T>(records: Model[], tableName: string): T[] => {
    // @ts-expect-error constructor.table not exposed in type definition
    return records.filter((r) => r.constructor.table === tableName) as T[];
};

export function extractChannelDisplayName(raw: Pick<Channel, 'type' | 'display_name'>, record?: ChannelModel) {
    let displayName = '';
    switch (raw.type) {
        case General.DM_CHANNEL:
            displayName = raw.display_name.trim() || record?.displayName || '';
            break;
        case General.GM_CHANNEL: {
            const rawMembers = raw.display_name.split(',').length;
            const recordMembers = record?.displayName.split(',').length || rawMembers;
            if (recordMembers < rawMembers && record?.displayName) {
                displayName = record.displayName;
            } else {
                displayName = raw.display_name;
            }
            break;
        }
        default:
            displayName = raw.display_name;
            break;
    }

    return displayName;
}
