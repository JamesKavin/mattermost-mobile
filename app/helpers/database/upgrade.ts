// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {getLastInstalledVersion} from '@queries/app/info';
import {logError, logInfo} from '@utils/log';

import type {DatabaseManager} from '@typings/database/manager';
import type InfoModel from '@typings/database/models/app/info';

export async function beforeUpgrade(serverUrls: string[], versionNumber: string, buildNumber: string) {
    const info = await getLastInstalledVersion();
    const manager: DatabaseManager | undefined = this.serverDatabases ? this : undefined;
    if (manager && serverUrls.length && info && (versionNumber !== info.versionNumber || buildNumber !== info.buildNumber)) {
        await beforeUpgradeTo450(manager, serverUrls, info);
    }
}

async function beforeUpgradeTo450(manager: DatabaseManager, serverUrls: string[], info: InfoModel) {
    try {
        const buildNumber = parseInt(info.buildNumber, 10);
        if (info.versionNumber === '2.0.0' && buildNumber < 450) {
            for await (const serverUrl of serverUrls) {
                logInfo('Remove database before upgrading for', serverUrl);
                await manager.deleteServerDatabaseFiles(serverUrl);
            }
        }
    } catch (e) {
        logError('Error running the upgrade before build 450', e);
    }
}
