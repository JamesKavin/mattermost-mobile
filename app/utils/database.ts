// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import urlParse from 'url-parse';

import DatabaseConnectionException from '@database/exceptions/database_connection_exception';
import DatabaseManager from '@database/manager';

type SetActiveDatabaseArgs = {
  serverUrl: string;
  displayName?: string;
};

export const createAndSetActiveDatabase = async ({serverUrl, displayName}: SetActiveDatabaseArgs) => {
    const connectionName = displayName ?? urlParse(serverUrl)?.hostname;

    try {
        const databaseClient = new DatabaseManager();
        await databaseClient.getDatabaseConnection({serverUrl, connectionName, setAsActiveDatabase: true});
    } catch (e) {
        throw new DatabaseConnectionException(
            `createAndSetActiveDatabase: Unable to create and set serverUrl ${serverUrl} as current active database with name ${displayName}`,
        );
    }
};

export const getDefaultDatabase = async () => {
    try {
        const databaseClient = new DatabaseManager();
        const defaultDatabase = await databaseClient.getDefaultDatabase();
        return {
            error: defaultDatabase ? null : 'Unable to retrieve the App database.',
            defaultDatabase,
        };
    } catch (e) {
        return {
            error: 'Unable to retrieve the App database.',
            defaultDatabase: null,
        };
    }
};

export const getActiveServerDatabase = async () => {
    try {
        const databaseClient = new DatabaseManager();
        const activeServerDatabase = await databaseClient.getActiveServerDatabase();

        return {
            error: activeServerDatabase ? null : 'Unable to retrieve the current active server database.',
            activeServerDatabase,
        };
    } catch (e) {
        return {
            error: 'Unable to retrieve the current active server database.',
            activeServerDatabase: null,
        };
    }
};
