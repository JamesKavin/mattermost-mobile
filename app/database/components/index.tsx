// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import React, {ComponentType, useState} from 'react';
import {Database} from '@nozbe/watermelondb';
import DatabaseProvider from '@nozbe/watermelondb/DatabaseProvider';

import {MM_TABLES} from '@constants/database';
import DatabaseManager from '@database/manager';
import Servers from '@app/database/models/app/servers';

const {SERVERS} = MM_TABLES.APP;

export function withServerDatabase<T>(
    Component: ComponentType<T>,
): ComponentType<T> {
    return function ServerDatabaseComponent(props) {
        const [database, setDatabase] = useState<Database|unknown>();
        const db = DatabaseManager.appDatabase?.database;

        const observer = async (servers: Servers[]) => {
            const server = servers.reduce((a, b) => (b.lastActiveAt > a.lastActiveAt ? b : a));

            const serverDatabase = DatabaseManager.serverDatabases[server.url].database;
            if (serverDatabase) {
                setDatabase(serverDatabase);
            }
        };

        db?.collections.
            get(SERVERS).
            query().
            observeWithColumns(['last_active_at']).
            subscribe(observer);

        if (!database) {
            return null;
        }

        return (
            <DatabaseProvider database={(database as Database)}>
                <Component {...props}/>
            </DatabaseProvider>
        );
    };
}
