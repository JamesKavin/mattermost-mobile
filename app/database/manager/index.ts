// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {Database, Q} from '@nozbe/watermelondb';
import SQLiteAdapter, {MigrationEvents} from '@nozbe/watermelondb/adapters/sqlite';
import logger from '@nozbe/watermelondb/utils/common/logger';
import {DeviceEventEmitter, Platform} from 'react-native';
import DeviceInfo from 'react-native-device-info';
import {FileSystem} from 'react-native-unimodules';
import urlParse from 'url-parse';

import {MIGRATION_EVENTS, MM_TABLES} from '@constants/database';
import AppDataOperator from '@database/operator/app_data_operator';
import AppDatabaseMigrations from '@app/database/migration/app';
import {Info, Global, Servers} from '@app/database/models/app';
import {schema as appSchema} from '@app/database/schema/app';
import ServerDatabaseMigrations from '@database/migration/server';
import {Channel, ChannelInfo, ChannelMembership, CustomEmoji, Draft, File,
    Group, GroupMembership, GroupsInChannel, GroupsInTeam, MyChannel, MyChannelSettings, MyTeam,
    Post, PostMetadata, PostsInChannel, PostsInThread, PreferenceModel, Reaction, Role,
    SlashCommand, SystemModel, Team, TeamChannelHistory, TeamMembership, TeamSearchHistory,
    TermsOfService, User,
} from '@database/models/server';
import {serverSchema} from '@database/schema/server';
import {queryActiveServer, queryServer} from '@queries/app/servers';
import {deleteIOSDatabase, getIOSAppGroupDetails} from '@utils/mattermost_managed';
import {hashCode} from '@utils/security';

import type {AppDatabase, CreateServerDatabaseArgs, RegisterServerDatabaseArgs, Models, ServerDatabase, ServerDatabases} from '@typings/database/database';
import {DatabaseType} from '@typings/database/enums';
import type IServers from '@typings/database/models/app/servers';

import ServerDataOperator from '../operator/server_data_operator';

const {SERVERS} = MM_TABLES.APP;
const APP_DATABASE = 'app';

class DatabaseManager {
  public appDatabase?: AppDatabase;
  public serverDatabases: ServerDatabases = {};
  private readonly appModels: Models;
  private readonly databaseDirectory: string | null;
  private readonly serverModels: Models;

  constructor() {
      this.appModels = [Info, Global, Servers];
      this.serverModels = [
          Channel, ChannelInfo, ChannelMembership, CustomEmoji, Draft, File,
          Group, GroupMembership, GroupsInChannel, GroupsInTeam, MyChannel, MyChannelSettings, MyTeam,
          Post, PostMetadata, PostsInChannel, PostsInThread, PreferenceModel, Reaction, Role,
          SlashCommand, SystemModel, Team, TeamChannelHistory, TeamMembership, TeamSearchHistory,
          TermsOfService, User,
      ];

      this.databaseDirectory = Platform.OS === 'ios' ? getIOSAppGroupDetails().appGroupDatabase : FileSystem.documentDirectory;
  }

  /**
  * init : Retrieves all the servers registered in the default database
  * @param {string[]} serverUrls
  * @returns {Promise<void>}
  */
  public init = async (serverUrls: string[]): Promise<void> => {
      await this.createAppDatabase();
      for await (const serverUrl of serverUrls) {
          await this.initServerDatabase(serverUrl);
      }
      this.appDatabase?.operator.handleInfo({
          info: [{
              build_number: DeviceInfo.getBuildNumber(),
              created_at: Date.now(),
              version_number: DeviceInfo.getVersion(),
          }],
          prepareRecordsOnly: false,
      });
  };

  /**
   * createAppDatabase: Creates the App database. However,
   * if a database could not be created, it will return undefined.
   * @returns {Promise<AppDatabase|undefined>}
   */
  private createAppDatabase = async (): Promise<AppDatabase|undefined> => {
      try {
          const databaseName = APP_DATABASE;

          const databaseFilePath = this.getDatabaseFilePath(databaseName);
          const modelClasses = this.appModels;
          const schema = appSchema;

          const adapter = new SQLiteAdapter({
              dbName: databaseFilePath,
              migrationEvents: this.buildMigrationCallbacks(databaseName),
              migrations: AppDatabaseMigrations,
              schema,
          });

          const database = new Database({adapter, actionsEnabled: true, modelClasses});
          const operator = new AppDataOperator(database);

          this.appDatabase = {
              database,
              operator,
          };

          return this.appDatabase;
      } catch (e) {
          // TODO : report to sentry? Show something on the UI ?
      }

      return undefined;
  };

  /**
   * createServerDatabase: Creates a server database and registers the the server in the app database. However,
   * if a database connection could not be created, it will return undefined.
   * @param {CreateServerDatabaseArgs} createServerDatabaseArgs
   *
   * @returns {Promise<ServerDatabase|undefined>}
   */
   public createServerDatabase = async ({config}: CreateServerDatabaseArgs): Promise<ServerDatabase|undefined> => {
       const {dbName, displayName, serverUrl} = config;

       if (serverUrl) {
           try {
               const databaseName = hashCode(serverUrl);
               const databaseFilePath = this.getDatabaseFilePath(databaseName);
               const migrations = ServerDatabaseMigrations;
               const modelClasses = this.serverModels;
               const schema = serverSchema;

               const adapter = new SQLiteAdapter({
                   dbName: databaseFilePath,
                   migrationEvents: this.buildMigrationCallbacks(databaseName),
                   migrations,
                   schema,
               });

               // Registers the new server connection into the DEFAULT database
               await this.addServerToAppDatabase({
                   databaseFilePath,
                   displayName: displayName || dbName,
                   serverUrl,
               });

               const database = new Database({adapter, actionsEnabled: true, modelClasses});
               const operator = new ServerDataOperator(database);
               const serverDatabase = {database, operator};

               this.serverDatabases[serverUrl] = serverDatabase;

               return serverDatabase;
           } catch (e) {
               // TODO : report to sentry? Show something on the UI ?
           }
       }

       return undefined;
   };

  /**
  * initServerDatabase : initializes the server database.
  * @param {string} serverUrl
  * @returns {Promise<void>}
  */
  private initServerDatabase = async (serverUrl: string): Promise<void> => {
      await this.createServerDatabase({
          config: {
              dbName: serverUrl,
              dbType: DatabaseType.SERVER,
              serverUrl,
          },
      });
  };

  /**
  * addServerToAppDatabase: Adds a record in the 'app' database - into the 'servers' table - for this new server connection
  * @param {string} databaseFilePath
  * @param {string} displayName
  * @param {string} serverUrl
  * @returns {Promise<void>}
  */
  private addServerToAppDatabase = async ({databaseFilePath, displayName, serverUrl}: RegisterServerDatabaseArgs): Promise<void> => {
      try {
          const isServerPresent = await this.isServerPresent(serverUrl); // TODO: Use normalized serverUrl

          if (this.appDatabase?.database && !isServerPresent) {
              const appDatabase = this.appDatabase.database;
              await appDatabase.action(async () => {
                  const serversCollection = appDatabase.collections.get(SERVERS);
                  await serversCollection.create((server: IServers) => {
                      server.dbPath = databaseFilePath;
                      server.displayName = displayName;
                      server.mentionCount = 0;
                      server.unreadCount = 0;
                      server.url = serverUrl; // TODO: Use normalized serverUrl
                      server.isSecured = urlParse(serverUrl).protocol === 'https';
                      server.lastActiveAt = 0;
                  });
              });
          }
      } catch (e) {
          // TODO : report to sentry? Show something on the UI ?
      }
  };

  /**
  * isServerPresent : Confirms if the current serverUrl does not already exist in the database
  * @param {String} serverUrl
  * @returns {Promise<boolean>}
  */
  private isServerPresent = async (serverUrl: string): Promise<boolean> => {
      if (this.appDatabase?.database) {
          const server = await queryServer(this.appDatabase.database, serverUrl);
          return Boolean(server);
      }

      return false;
  }

  /**
   * getActiveServerUrl: Get the record for active server database.
   * @returns {Promise<string|null|undefined>}
   */
  public getActiveServerUrl = async (): Promise<string|null|undefined> => {
      const database = this.appDatabase?.database;
      if (database) {
          const server = await queryActiveServer(database);
          return server?.url;
      }

      return null;
  }

  /**
   * getActiveServerDatabase: Get the record for active server database.
   * @returns {Promise<Database|undefined>}
   */
   public getActiveServerDatabase = async (): Promise<Database|undefined> => {
       const database = this.appDatabase?.database;
       if (database) {
           const server = await queryActiveServer(database);
           if (server?.url) {
               return this.serverDatabases[server.url].database;
           }
       }

       return undefined;
   }

  /**
   * setActiveServerDatabase: Set the new active server database.
   * This method should be called when switching to another server.
   * @param {string} serverUrl
   * @returns {Promise<void>}
   */
  public setActiveServerDatabase = async (serverUrl: string): Promise<void> => {
      if (this.appDatabase?.database) {
          const database = this.appDatabase?.database;
          await database.action(async () => {
              const servers = await database.collections.get(SERVERS).query(Q.where('url', serverUrl)).fetch();
              if (servers.length) {
                  servers[0].update((server: Servers) => {
                      server.lastActiveAt = Date.now();
                  });
              }
          });
      }
  };

  /**
   * deleteServerDatabase: Removes the *.db file from the App-Group directory for iOS or the files directory on Android.
   * Also, it sets the last_active_at to '0' entry in the 'servers' table from the APP database
   * @param  {string} serverUrl
   * @returns {Promise<boolean>}
   */
  public deleteServerDatabase = async (serverUrl: string): Promise<void> => {
      if (this.appDatabase?.database) {
          const database = this.appDatabase?.database;
          const server = await queryServer(database, serverUrl);
          if (server) {
              database.action(async () => {
                  await server.update((record) => {
                      record.lastActiveAt = 0;
                  });
              });

              delete this.serverDatabases[serverUrl];
              this.deleteServerDatabaseFiles(serverUrl);
          }
      }
  }

  /**
   * destroyServerDatabase: Removes the *.db file from the App-Group directory for iOS or the files directory on Android.
   * Also, removes the entry in the 'servers' table from the APP database
   * @param  {string} serverUrl
   * @returns {Promise<boolean>}
   */
  public destroyServerDatabase = async (serverUrl: string): Promise<void> => {
      if (this.appDatabase?.database) {
          const database = this.appDatabase?.database;
          const server = await queryServer(database, serverUrl);
          if (server) {
              database.action(async () => {
                  await server.destroyPermanently();
              });

              delete this.serverDatabases[serverUrl];
              this.deleteServerDatabaseFiles(serverUrl);
          }
      }
  }

  /**
   * deleteServerDatabaseFiles: Removes the *.db file from the App-Group directory for iOS or the files directory on Android.
   * @param  {string} serverUrl
   * @returns {Promise<void>}
   */
  private deleteServerDatabaseFiles = async (serverUrl: string): Promise<void> => {
      const databaseName = hashCode(serverUrl);

      if (Platform.OS === 'ios') {
          // On iOS, we'll delete the *.db file under the shared app-group/databases folder
          deleteIOSDatabase({databaseName});
          return;
      }

      // On Android, we'll delete both the *.db file and the *.db-journal file
      const androidFilesDir = `${this.databaseDirectory}databases/`;
      const databaseFile = `${androidFilesDir}${databaseName}.db`;
      const databaseJournal = `${androidFilesDir}${databaseName}.db-journal`;

      await FileSystem.deleteAsync(databaseFile);
      await FileSystem.deleteAsync(databaseJournal);
  }

  /**
   * factoryReset: Removes the databases directory and all its contents on the respective platform
   * @param {boolean} shouldRemoveDirectory
   * @returns {Promise<boolean>}
   */
  factoryReset = async (shouldRemoveDirectory: boolean): Promise<boolean> => {
      try {
          //On iOS, we'll delete the databases folder under the shared AppGroup folder
          if (Platform.OS === 'ios') {
              deleteIOSDatabase({shouldRemoveDirectory});
              return true;
          }

          // On Android, we'll remove the databases folder under the Document Directory
          const androidFilesDir = `${this.databaseDirectory}databases/`;
          await FileSystem.deleteAsync(androidFilesDir);
          return true;
      } catch (e) {
          return false;
      }
  };

  /**
   * buildMigrationCallbacks: Creates a set of callbacks that can be used to monitor the migration process.
   * For example, we can display a processing spinner while we have a migration going on. Moreover, we can also
   * hook into those callbacks to assess how many of our servers successfully completed their migration.
   * @param {string} dbName
   * @returns {MigrationEvents}
   */
  private buildMigrationCallbacks = (dbName: string) => {
      const migrationEvents: MigrationEvents = {
          onSuccess: () => {
              return DeviceEventEmitter.emit(MIGRATION_EVENTS.MIGRATION_SUCCESS, {
                  dbName,
              });
          },
          onStart: () => {
              return DeviceEventEmitter.emit(MIGRATION_EVENTS.MIGRATION_STARTED, {
                  dbName,
              });
          },
          onError: (error: Error) => {
              return DeviceEventEmitter.emit(MIGRATION_EVENTS.MIGRATION_ERROR, {
                  dbName,
                  error,
              });
          },
      };

      return migrationEvents;
  };

  /**
   * getDatabaseFilePath: Using the database name, this method will return the database file path for each platform.
   * On iOS, it will point towards the AppGroup shared directory while on Android, it will point towards the Files Directory.
   * Please note that in each case, the *.db files will be created/grouped under a 'databases' sub-folder.
   * iOS Simulator : appGroup => /Users/{username}/Library/Developer/CoreSimulator/Devices/DA6F1C73/data/Containers/Shared/AppGroup/ACA65327/databases"}
   * Android Device: file:///data/user/0/com.mattermost.rnbeta/files/databases
   *
   * @param {string} dbName
   * @returns {string}
   */
  private getDatabaseFilePath = (dbName: string): string => {
      return Platform.OS === 'ios' ? `${this.databaseDirectory}/${dbName}.db` : `${this.databaseDirectory}${dbName}.db`;
  };
}

if (!__DEV__) {
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
    logger.silence();
}

export default new DatabaseManager();
