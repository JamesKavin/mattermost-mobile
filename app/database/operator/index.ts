// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import BaseHandler, {BaseHandlerMix} from '@database/operator/handlers/base_handler';
import ChannelHandler, {ChannelHandlerMix} from '@database/operator/handlers/channel';
import GroupHandler, {GroupHandlerMix} from '@database/operator/handlers/group';
import PostHandler, {PostHandlerMix} from '@database/operator/handlers/post';
import TeamHandler, {TeamHandlerMix} from '@database/operator/handlers/team';
import UserHandler, {UserHandlerMix} from '@database/operator/handlers/user';
import {DatabaseInstance} from '@typings/database/database';
import mix from '@utils/mix';

interface Operator extends BaseHandlerMix, PostHandlerMix, UserHandlerMix, GroupHandlerMix, ChannelHandlerMix, TeamHandlerMix {}

class Operator extends mix(BaseHandler).with(PostHandler, UserHandler, GroupHandler, ChannelHandler, TeamHandler) {
    database: DatabaseInstance;

    constructor(activeDatabase?: DatabaseInstance) {
        super();
        if (activeDatabase) {
            this.activeDatabase = activeDatabase;
        }
    }
}

export default Operator;
