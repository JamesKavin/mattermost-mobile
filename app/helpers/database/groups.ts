// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {MM_TABLES} from '@constants/database';

import type ChannelModel from '@typings/database/models/servers/channel';
import type GroupModel from '@typings/database/models/servers/group';
import type GroupsChannelModel from '@typings/database/models/servers/groups_channel';
import type GroupsTeamModel from '@typings/database/models/servers/groups_team';
import type GroupMembershipModel from '@typings/database/models/servers/group_membership';
import type PostModel from '@typings/database/models/servers/post';
import type TeamModel from '@typings/database/models/servers/team';

export const queryGroupsAssociatedToChannelForReference = async (channel: ChannelModel) => {
    const groupChannels = await channel.groupsChannel.fetch() as GroupsChannelModel[];
    const groupChannelPromises = groupChannels.map((g) => g.group.fetch());
    const groups = await Promise.all(groupChannelPromises) as GroupModel[];
    return groups.filter((g) => g.deleteAt === 0);
};

export const queryGroupsAssociatedToTeamForReference = async (team: TeamModel) => {
    const groupTeams = await team.groupsTeam.fetch() as GroupsTeamModel[];
    const teamGroupPromises = groupTeams.map((g) => g.group.fetch());
    const groups = await Promise.all(teamGroupPromises) as GroupModel[];
    return groups.filter((g) => g.deleteAt === 0);
};

export const queryGroupForPosts = async (post: PostModel) => {
    try {
        if (post.id === post.pendingPostId) {
            return null;
        }

        const channel = await post.channel.fetch() as ChannelModel;
        const team = await channel?.team.fetch() as TeamModel | undefined;
        let groupsForReference: GroupModel[] | null = null;

        if (team?.isGroupConstrained && channel.isGroupConstrained) {
            const groupsFromChannel = await queryGroupsAssociatedToChannelForReference(channel);
            const groupsFromTeam = await queryGroupsAssociatedToTeamForReference(team);
            groupsForReference = groupsFromChannel.concat(groupsFromTeam.filter((item) => groupsFromChannel.indexOf(item) < 0));
        } else if (team?.isGroupConstrained) {
            groupsForReference = await queryGroupsAssociatedToTeamForReference(team);
        } else if (channel.isGroupConstrained) {
            groupsForReference = await queryGroupsAssociatedToChannelForReference(channel);
        } else {
            const myGroups = await post.collections.get(MM_TABLES.SERVER.GROUP_MEMBERSHIP).query().fetch() as GroupMembershipModel[];
            const myGroupsPromises = myGroups.map((g) => g.group.fetch());
            groupsForReference = await Promise.all(myGroupsPromises) as GroupModel[];
        }

        return groupsForReference;
    } catch {
        return null;
    }
};
