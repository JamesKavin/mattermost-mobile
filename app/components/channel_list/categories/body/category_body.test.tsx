// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {Database, Q} from '@nozbe/watermelondb';
import React from 'react';

import {MM_TABLES} from '@constants/database';
import {renderWithEverything} from '@test/intl-test-helper';
import TestHelper from '@test/test_helper';

import CategoryBody from './category_body';

import type CategoryModel from '@typings/database/models/servers/category';
import type CategoryChannelModel from '@typings/database/models/servers/category_channel';
import type ChannelModel from '@typings/database/models/servers/channel';
import type MyChannelModel from '@typings/database/models/servers/my_channel';

const {SERVER: {CATEGORY, CATEGORY_CHANNEL, CHANNEL, MY_CHANNEL}} = MM_TABLES;

describe('components/channel_list/categories/body', () => {
    let database: Database;
    let category: CategoryModel;
    let categoryChannels: CategoryChannelModel[];
    let myChannels: MyChannelModel[];
    let channels: ChannelModel[];

    beforeAll(async () => {
        const server = await TestHelper.setupServerDatabase();
        database = server.database;

        const categories = await database.get<CategoryModel>(CATEGORY).query(
            Q.take(1),
        ).fetch();

        category = categories[0];

        categoryChannels = await database.get<CategoryChannelModel>(CATEGORY_CHANNEL).query(
            Q.where('category_id', category.id),
        ).fetch();

        const channelIds = await database.get<ChannelModel>(CHANNEL).query(
            Q.on(CATEGORY_CHANNEL, 'category_id', category.id),
        ).fetchIds();

        myChannels = await database.get<MyChannelModel>(MY_CHANNEL).query(
            Q.where('id', Q.oneOf(channelIds)),
        ).fetch();
    });

    it('should match snapshot', () => {
        const wrapper = renderWithEverything(
            <CategoryBody
                category={category}
                myChannels={myChannels}
                categoryChannels={categoryChannels}
                channels={channels}
            />,
            {database},
        );
        expect(wrapper.toJSON()).toMatchSnapshot({
            props: {data: expect.anything()},
        });
    });
});
