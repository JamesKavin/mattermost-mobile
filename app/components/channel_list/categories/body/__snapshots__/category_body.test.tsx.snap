// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/channel_list/categories/body should match snapshot 1`] = `
Object {
  "children": Array [
    <View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          animatedStyle={
            Object {
              "value": Object {
                "height": 40,
                "marginVertical": 2,
                "opacity": 1,
              },
            }
          }
          collapsable={false}
          style={
            Object {
              "height": 40,
              "marginVertical": 2,
              "opacity": 1,
            }
          }
        >
          <View
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "opacity": 1,
              }
            }
          >
            <View
              style={
                Object {
                  "alignItems": "center",
                  "flexDirection": "row",
                  "height": 40,
                  "paddingLeft": 2,
                }
              }
              testID="category.test_category.channel_list_item.channel.collapsed.false"
            >
              <View
                style={
                  Array [
                    Object {
                      "alignItems": "center",
                      "justifyContent": "center",
                    },
                    Object {
                      "height": 24,
                      "width": 24,
                    },
                    undefined,
                    undefined,
                  ]
                }
              >
                <Icon
                  name="globe"
                  style={
                    Array [
                      Object {
                        "color": "rgba(255,255,255,0.4)",
                      },
                      undefined,
                      undefined,
                      Object {
                        "fontSize": 24,
                        "left": 1,
                      },
                    ]
                  }
                  testID="undefined.public"
                />
              </View>
              <Text
                ellipsizeMode="tail"
                numberOfLines={1}
                style={
                  Array [
                    Object {
                      "fontFamily": "OpenSans",
                      "fontSize": 16,
                      "fontWeight": "400",
                      "lineHeight": 24,
                    },
                    Object {
                      "color": "rgba(255,255,255,0.72)",
                      "flex": 1,
                      "marginTop": -1,
                      "paddingLeft": 12,
                    },
                    false,
                    false,
                  ]
                }
                testID="category.test_category.channel_list_item.channel.display_name"
              >
                Channel
              </Text>
            </View>
          </View>
        </View>
      </View>
    </View>,
  ],
  "props": Object {
    "data": Anything,
    "getItem": [Function],
    "getItemCount": [Function],
    "initialNumToRender": 20,
    "invertStickyHeaders": undefined,
    "keyExtractor": [Function],
    "onContentSizeChange": [Function],
    "onLayout": [Function],
    "onMomentumScrollBegin": [Function],
    "onMomentumScrollEnd": [Function],
    "onScroll": [Function],
    "onScrollBeginDrag": [Function],
    "onScrollEndDrag": [Function],
    "removeClippedSubviews": true,
    "renderItem": [Function],
    "scrollEventThrottle": 50,
    "stickyHeaderIndices": Array [],
    "style": undefined,
    "updateCellsBatchingPeriod": 10,
    "viewabilityConfigCallbackPairs": Array [],
    "windowSize": 15,
  },
  "type": "RCTScrollView",
}
`;
